Port MASTG-TEST-0028: Testing for Testing Deep Links (android) (by @appknox)

ScreaMy7 opened the pull request:
closes #2980

ScreaMy7 committed:
added testing deeplinks

ScreaMy7 committed:
after review

ScreaMy7 committed:
md fix

ScreaMy7 committed:
moved to platform

cpholguera committed:
Merge branch 'master' into Testing-deeplink-draft

cpholguera commented on tests-beta/android/MASVS-PLATFORM/MASTG-TEST-0292.md:
```diff
@@ -0,0 +1,24 @@
+platform: android
+id: MASTG-TEST-0292
+type: [static]
+weakness: MASWE-0088
```
the right MASWE should be [MASWE-0071: WebViews Loading Content from Untrusted Sources
](https://mas.owasp.org/MASWE/MASVS-PLATFORM/MASWE-0071/)

```suggestion
weakness: MASWE-0071
```

cpholguera commented on tests/android/MASVS-PLATFORM/MASTG-TEST-0028.md:
@ScreaMy7 ⚠️ **PLEASE BE CAREFUL:** 

I see that a lot of the information contained in this test is being lost right now. Please consider porting it all:

- What information is **knowledge** and belongs in https://mas.owasp.org/MASTG/knowledge/android/MASVS-PLATFORM/MASTG-KNOW-0019/
- What information describes **best practices** and should be added to a new `MASTG-BEST-XXXX`? (each test should have at least one best practice linked)
- What **techniques** are here that don't have a MASTG-TECH-XXXX entry? e.g. 
  - "Check for Deep Link Usage"
  - "Check for Correct Website Association"
  - "Monitoring Deep Links"
  - "Invoking Deep Links"
  - etc
- What **tools** are used and don't have a MASTG-TOOL-XXXX entry? e.g. "Android App Link Verification Tester"
- What **tests** are missing?
  - "Use of Unverified App Links"
  - "Use of Unverified Custom URL Schemes"
  - "Use of Verified App Links with Incorrect Website Association"

cpholguera commented on tests-beta/android/MASVS-PLATFORM/MASTG-TEST-0293.md:
```diff
@@ -0,0 +1,24 @@
+---
+title: Deep Link Intent Filter Missing android:autoVerify
```
This should be "Use of Unverified App Links" plus an additional "Use of Unverified Custom URL Schemes" which is now missing.

cpholguera commented:
Also please use "fake IDs" (e.g. MASTG-TEST-0x28-1, MASTG-TEST-0x28-2, MASTG-DEMO-00xx-1, MASTG-DEMO-00xx-2) until we're ready to merge, to avoid conflicts.

ScreaMy7 committed:
add new tech and demos.

cpholguera committed:
Merge branch 'master' into Testing-deeplink-draft

ScreaMy7 committed:
added fix and original manifest

cpholguera committed:
Merge branch 'master' into Testing-deeplink-draft

ScreaMy7 commented on tests-beta/android/MASVS-PLATFORM/MASTG-TEST-XXXA.md:
```diff
@@ -0,0 +1,25 @@
+type: [static]
+weakness: MASWE-0071
+profiles: [L1, L2]
+knowledge: [MASTG-KNOW-0019]
```
MASTG know 0019 exist.

cpholguera committed:
Merge branch 'master' into Testing-deeplink-draft

ScreaMy7 committed:
added requested changes

serek8 reviewed (commented):
Thanks for contributing! I have a couple of remarks:
* Can you link MASTG-KNOW-0019 knowledge page and add some missing parts from the old test:
  * the change in Android 12 that impacts non-verifiable (We had it in the old [test](https://mas.owasp.org/MASTG/tests/android/MASVS-PLATFORM/MASTG-TEST-0028/))
  * the reports from the hackerone, they are very illustrative
* Can you add a Best Practise page and guide the user to use App Links and explain why
 
Additionally: Not a must for this PR. We can create a new issue for it. Currently we only test receiving the deep links in the app. I think we could also create some best practices on how to invoke app links. I didn't give it much thought yet but maybe the payload could be encrytpted so it can't leak sensitive data and only the legit app/website can decrypt it. Or maybe we can somehow else ensure that the URL won't be hijacked by another app and goes directly to the expected app or website?

ScreaMy7 committed:
added best practise for app links

serek8 commented on tests-beta/android/MASVS-PLATFORM/MASTG-TEST-0x02.md:
```diff
@@ -0,0 +1,35 @@
+
+The Android version the app runs on also influences the risk. Before Android 12 (API level 31), if the app has any [non-verifiable links](https://developer.android.com/training/app-links/verify-android-applinks#fix-errors) (e.g., missing `autoVerify`, invalid Digital Asset Links file, or custom URL schemes), the system may skip verification for all Android App Links declared by that app—leaving e ...
+
+This vulnerability occurs when a deep link intent filter in `AndroidManifest.xml` lacks the `android:autoVerify="true"` attribute. Without verification, Android cannot confirm the app's ownership of the declared domain. A malicious app could register the same intent filter and intercept deep links, enabling phishing, credential theft, or hijacking of user actions.
```
I am not an expert in App Links but this got me curious:

> A malicious app could register the same intent filter and intercept deep links, enabling phishing, credential theft, or hijacking of user actions.

So on Android 11, if my legit app has `autoVerify="true"` but my server returns 404 for the verification, my app won't receive events for this link. But a malicious app with `autoVerify="false"` will receive these events instead? Or you mean that the system will open "Open with..." dialog and a user can choose a malicious app there? Or both scenarios are possible?

ScreaMy7 replied:
When a legitimate app has autoVerify="true" but verification fails (404), that domain becomes "unverified" for your app.  So it falls back to acting like a regular deep link handle. A malicious app with
  autoVerify="false" can then intercept those links because both apps are now considered equally "unverified" for that domain. I tested this scenario on Android 12, I received the "Open with" dialog box.

serek8 commented on tests-beta/android/MASVS-PLATFORM/MASTG-TEST-XXXC.md:
```diff
@@ -0,0 +1,25 @@
+---
+title: Use of Unverified Custom URL Schemes
```
Aren't `Custom URL Schemes` always `Unverified`? Only AppLinks can be `Verified` / `Non Verified`, no? I believe we want to detect use of any `Custom URL Schemes` because they are fundamentally insecure?

cpholguera replied:
That's correct.

serek8 replied:
Then maybe we should name this test "Use of Custom URL Schemes"?

serek8 commented on tests-beta/android/MASVS-PLATFORM/MASTG-TEST-XXXC.md:
```diff
@@ -0,0 +1,25 @@
+---
+title: Use of Unverified Custom URL Schemes
```
```suggestion
title: Use of Custom URL Schemes
```

ScreaMy7 committed:
title change as suggested

serek8 commented on rules/mastg-android-autoverify-missing.yml:
```diff
@@ -0,0 +1,24 @@
+          <action android:name="android.intent.action.VIEW" />
+          <category android:name="android.intent.category.DEFAULT" />
+          <category android:name="android.intent.category.BROWSABLE" />
+          <data android:scheme="https" />
```
Should we also cover `<data android:scheme="http" />`?

ScreaMy7 replied:
Yes, app Links apply to both http and https. I'll update the rule to catch both schemes.

serek8 commented on rules/mastg-android-custom-deeplink-scheme.yml:
```diff
@@ -0,0 +1,11 @@
+rules:
+- id: mastg-android-custom-deeplink-intent-filter
```
ID doesn't match filename
```suggestion
- id: mastg-android-custom-deeplink-scheme
```

serek8 commented on demos/android/MASVS-PLATFORM/MASTG-DEMO-XXXC/output.txt:
```diff
@@ -0,0 +1,20 @@
+└────────────────┘
+                                               
+    AndroidManifest_reversed.xml
+    ❯❱ android-custom-deeplink-intent-filter
```
ID doesn't match
```suggestion
    ❯❱ mastg-android-custom-deeplink-scheme
```

serek8 commented on rules/mastg-android-autoverify-missing.yml:
```diff
@@ -0,0 +1,24 @@
+          <category android:name="android.intent.category.DEFAULT" />
+          <category android:name="android.intent.category.BROWSABLE" />
+          <data android:scheme="..." />
+        </intent-filter>
```
Should we worry about this one @ScreaMy7

serek8 reviewed (commented):
I briefly had a look and found cosmetic some issues. I still need to run the DEMOs myself

ScreaMy7 committed:
added http to demo and updated files

cpholguera committed:
Merge branch 'master' of https://github.com/OWASP/mastg into pr/ScreaMy7/3463

cpholguera committed:
use fake IDs

cpholguera commented on demos/android/MASVS-PLATFORM/MASTG-DEMO-0x01/MastgTest.kt:
This demo doesn't work. Nothing happens when running the adb command.

ScreaMy7 replied:
Let me check.

ScreaMy7 replied:
Fixed.

cpholguera commented on demos/android/MASVS-PLATFORM/MASTG-DEMO-0x02/MASTG-DEMO-0x02.md:
Demos must be fully functional. In this case, the link should perform an action so that we can demonstrate how any app can trigger that sensitive action due to a lack of verification.

ScreaMy7 replied:
Addressed. The demo was manifest-only with no runtime action so I have added App Link and that performs a sensitive action.

ScreaMy7 committed:
Make MASTG-DEMO-0x02 functional: trigger a sensitive action via the unverified App Link

Addresses review feedback (r3448625913): demos must be fully functional,
and the link should perform an action demonstrating how an unverified App
Link can be triggered by any app.

Previously DEMO-0x02 was manifest-only: it documented the missing
android:autoVerify but never demonstrated the consequence, and the filter
sat on MainActivity (which ignores the launch intent), so the link did
nothing at runtime.

Add MastgTest.kt with an exported DeepLinkActivity that handles the
http(s)://deeplink.example.com App Link (registered WITHOUT
android:autoVerify) and performs a sensitive action - disabling two-factor
authentication - with no verification or user confirmation. The App Link
filter is moved from MainActivity to DeepLinkActivity; MainActivity keeps
the LAUNCHER filter.

Regenerated AndroidManifest_reversed.xml and MastgTest_reversed.java via
tools/extract-code-for-mastg-demo.sh and refreshed output.txt by running
run.sh; the autoVerify-missing semgrep finding still holds. Documented the
adb trigger command in the Evaluation.

Verified on a physical device: opening the link stores
two_factor_enabled=false (set_by=deeplink), confirming the sensitive
action fires.

ScreaMy7 committed:
Fix MASTG-DEMO-0x01 deep link not triggering at runtime

The demo's vulnerable sink processDeepLinkAndLoad() was never invoked:
the vulnerable-app://deeplink VIEW filter sat on MainActivity, which
only reads intent data via the Start button and ignores the launch
intent, so `adb shell am start ... VIEW` did nothing.

Route the deep link to a dedicated exported DeepLinkActivity whose
onCreate reads intent.data and calls processDeepLinkAndLoad(), loading
the attacker-controlled url into the WebView. Verified on a physical
device: the adb command now renders the URL in the WebView.

Regenerated MastgTest_reversed.java and AndroidManifest_reversed.xml via
tools/extract-code-for-mastg-demo.sh. The getQueryParameter -> loadUrl
data flow is unchanged, so the semgrep finding in output.txt still holds.

cpholguera commented on rules/mastg-android-autoverify-missing.yml:
This rule appears to be too brittle for MASTG coverage. It assumes a specific child order inside the intent filter, VIEW, then DEFAULT, then BROWSABLE, then data. That means it can miss valid manifests where the same elements are present but ordered differently. It also requires DEFAULT, so any app that omits that category is excluded entirely.

Since MASTG rules are meant to work against arbitrary real world apps, the rule should not depend on manifest element ordering unless ordering is semantically required. In this case, the current pattern can produce false negatives for structurally equivalent intent filters, which goes against the intended coverage goals.

cpholguera commented on rules/mastg-android-custom-deeplink-scheme.yml:
This rule is brittle because it relies on a raw regex over the manifest text rather than XML semantics. It only matches a literal `<intent-filter>` opening tag, so it can miss valid filters with attributes, whitespace, or formatting differences, such as `<intent-filter android:autoVerify="true">` or `<intent-filter >`.

It also only matches double quoted `android:scheme` values, even though XML allows single quoted attributes. Since the rule is not XML aware, it can also match text in comments or otherwise behave unexpectedly across formatting edge cases.
